Modules/Registration/Common/test/itkMeanSquaresImageMetricTest.cxx

```diff
@@ -295,6 +298,110 @@ itkMeanSquaresImageMetricTest(int, char *[])
   //-------------------------------------------------------
   metric->Print(std::cout);
 
+
+   //-------------------------------------------------------
+  // Check consistency between BSplineTransform
+  // derivatives computed with vs omitting weights caching.
+  //-------------------------------------------------------
+
+  // Define types/constants
+  constexpr unsigned int splineOrder = 3;
+  constexpr unsigned int nodesPerDimension = 8;
+  using BSplineTransformType = itk::BSplineTransform<CoordinateRepresentationType, ImageDimension, splineOrder>;
+  using InitializerType = itk::BSplineTransformInitializer<BSplineTransformType, MovingImageType>;
+
+  // Offset moving image to ensure derivatives for BSplineTransform are non-zero.
+  MovingImageType::PointValueType bSplineTestMovingImageOrigin[] = { 1.0f, 1.0f };
+  movingImage->SetOrigin(bSplineTestMovingImageOrigin);
+
+  // Initialise BSplineTransform
+  auto bSplineTransform = BSplineTransformType::New();
+  auto initializer = InitializerType::New();
+  BSplineTransformType::MeshSizeType meshSize;
+  meshSize.Fill(nodesPerDimension - splineOrder);
+  initializer->SetTransform(bSplineTransform);
+  initializer->SetImage(movingImage);
+  initializer->SetTransformDomainMeshSize(meshSize);
+  initializer->InitializeTransform();
+
+  // Set bSplineTransform parameters
+  ParametersType bSplineParameters(bSplineTransform->GetNumberOfParameters());
+  bSplineParameters.Fill(0.0);
+  bSplineTransform->SetParameters(bSplineParameters);
+
+  // Initialise metric without caching
+  auto metricNoCaching = MetricType::New();
+  metricNoCaching->SetFixedImage(fixedImage);
+  metricNoCaching->SetMovingImage(movingImage);
+  metricNoCaching->SetInterpolator(interpolator);
+  metricNoCaching->SetFixedImageRegion(fixedImage->GetBufferedRegion());
+  metricNoCaching->SetTransform(bSplineTransform);
+  metricNoCaching->SetUseCachingOfBSplineWeights(false);
+  try
+  {
+    metricNoCaching->Initialize();
+  }
+  catch (const itk::ExceptionObject & e)
+  {
+    std::cout << "Failed to initialise metric to test BSplineTransform derivative computation omitting weights caching."
+              << std::endl;
+    std::cout << "Reason " << e.GetDescription() << std::endl;
+
+    return EXIT_FAILURE;
+  }
+
+  // Initialise metric using caching
+  auto metricWithCaching = MetricType::New();
```

Comment:
Why not reuse variable `metricNoCaching, e.g.:
```cpp
MetricType::DerivativeType derivativeNoCaching;
metricNoCaching->GetDerivative(bSplineParameters, derivativeNoCaching);

metricNoCaching->SetUseCachingOfBSplineWeights(true);
MetricType::DerivativeType derivativeWithCaching;
metricNoCaching->GetDerivative(bSplineParameters, derivativeWithCaching);
```
And rename it in that case to e.g. `metric`. Having two variables would make sense if we evaluated derivates many times.

Reply:
I may have been overly cautious. I wanted to be certain that the derivatives being set by reference were different, not just pointing to the same buffer. Something along the lines of trying to be correct and redundant rather than efficient but wrong.

Reply:
Not sure whether I'll finish revising code changes today - I'm planning to randomly initialise each parameter with MersenneTwister. Will that be ok?

Reply:
Code being revised to use single metric object to test caching vs no caching.